drv/AbstractMap.drv

```diff
@@ -67,7 +67,22 @@ public abstract class ABSTRACT_MAP KEY_VALUE_GENERIC extends ABSTRACT_FUNCTION K
 	public boolean isEmpty() {
 		return size() == 0;
 	}
-
+		
+#if VALUES_PRIMITIVE
+	/** Adds an increment to value currently associated with a key.
+	 *
+	 * <p>Note that this method respects the {@linkplain #defaultReturnValue() default return value} semantics: when
+	 * called with a key that does not currently appears in the map, the key
+	 * will be associated with the default return value plus
+	 * the given increment.
+	 *
+	 * @param k the key.
+	 * @param incr the increment.
+	 * @return the old value, or the {@linkplain #defaultReturnValue() default return value} if no value was present for the given key.
+	 */
+	public VALUE_GENERIC_TYPE addTo(final KEY_GENERIC_TYPE k, final VALUE_GENERIC_TYPE incr) { throw new UnsupportedOperationException(); }
```

Comment:
Isn't this a special case of merge / compute?

Reply:
Kinda, but this is the object allocation free variant and only works with numeric values since it is counting, and if it is not present then it just adds it with 0 delay.

Compute/Merge is basically slower due to the lambda.

Reply:
Overhead of e.g., `Double::sum` vs `a + b` very likely is negligible I'd guess. However, I think there aren't primitive binary operators so there would be boxing...

Reply:
Eh fastUtil add comparators themselves.

Reply:
Just to give some perspective @incaseoftrouble:
![image](https://user-images.githubusercontent.com/2556269/94617761-6bba4f00-02aa-11eb-8514-7e81a24433a8.png)
That is the compare code and the results from a JUnit4 Benchmark are these:
```
MiscTest.compareList: [measured 1000 out of 1100 rounds, threads: 1 (sequential)]
 round: 0.01 [+- 0.00], round.block: 0.00 [+- 0.00], round.gc: 0.00 [+- 0.00], GC.calls: 14, GC.time: 0.01, time.total: 11.26, time.warmup: 1.13, time.bench: 10.14

MiscTest.compareListLambda: [measured 1000 out of 1100 rounds, threads: 1 (sequential)]
 round: 0.01 [+- 0.00], round.block: 0.00 [+- 0.00], round.gc: 0.00 [+- 0.00], GC.calls: 19, GC.time: 0.01, time.total: 12.50, time.warmup: 1.23, time.bench: 11.28
```
It is only a tiny differences but some factors you can see is: GC is running more and the code runs a tiny bit slower.
instead of 10,14ms per execution it takes a full 1,14ms longer to execute the same code with the same seed just because a lambda was added. Rule of thumb is, the less a lambda is used or created the better it is. That is why I rather promote non lambda solutions.

Reply:
Yes, it is a 10% difference (even though I think the setup of the test should be separated), which is even less then I expected (its barely significant). Of course lambdas come with a cost. But thats not why lambdas were introduced in the first place.

From a practical perspective: Its a 1ms difference on a list with ~100k elements (so something like 1 million calls, assuming `n log n`). So, this hardly makes a difference (if you are dealing with anything where a difference of 1ms for 1 million calls makes a difference you should probably be writing in assembler). Maintaining an additional function with special semantics is a real difference. You could probably spend time in *any* other part of this library and optimize a bit to obtain a more noticeable effect. In my eyes, this is the prototical "premuature optimization" Knuth warned about.

Reply:
Btw AFAIK method refence lambdas are handled differently than anonymous classes. So `List.sort(Integer::compare)` should be noticeably faster than `List.sort((a, b) -> b - a)`

Reply:
Note right out of the gate: I know why lambdas exist, and I see their use cases and use them myself.
But they should never be the priority.
About the performance gain. You are aware that Integer::compare is just ((k, v) -> Integer.compare(k, v)) it is still a allocation.
It can be compared to a += b or a = a + b;
And since lambdas are rarely cache-able (speaking from experience) and the 10% lag comes from how often they are used in a single allocation what do you think does this to performance if you don't have a 1-10ms task, because you are reallocation the same function every time it is called.
So overhead rises with less executions per call.

Simple sum up from me: Lambdas are useful if performance isn't a factor, UI interaction, rarely called lazyloads, or some other cases. But otherwise avoid using them, because you could run a lot more code otherwise.

Also I hope I do not sound rude, i find this discussion really interesting.

Reply:
Also why i added this "addTo" function to the interface was so you could use object2IntMap instead of Object2ObjectOpenHashMap to access that function. Just to remove unnecessary casts or forced breaking of the rule that you should use interfaces on the left.

Reply:
>  About the performance gain. You are aware that Integer::compare is just ((k, v) -> Integer.compare(k, v)) it is still a allocation.

Technically, these two are quite different under the hood. The JVM can optimize direct method references much easier (see, e.g. [here](https://softwareengineering.stackexchange.com/questions/277473/is-there-a-performance-benefit-to-using-the-method-reference-syntax-instead-of-l)).

> It can be compared to a += b or a = a + b;

These two are the same (see e.g. [here](https://stackoverflow.com/questions/39821936/performance-of-in-java)). Compilers are incredibly smart.

>  So overhead rises with less executions per call.

Yes and no. The *absolute* overhead of a lambda is minimal.

> Simple sum up from me: Lambdas are useful if performance isn't a factor, UI interaction, rarely called lazyloads, or some other cases. But otherwise avoid using them, because you could run a lot more code otherwise.

They are also useful for core performance areas, called hundreds of times per second. Simply because you save so much effort and make code maintainable so that you can spend effort on where it actually matters. Just see the example, even though there are 1 million (!) calls to the lambda, the execution takes a few milliseconds. I'd bet money that in basically every single application where the `addTo` method is used, a (primitive) lambda together with some optimization elsewhere has a larger impact.

Given that you have infinite time, its of course most performant to not use lambdas at all and hand tailor every single datastructure. However, usually you don't have infinite time, so prioritizing clean, readable code actually yields more performant code in practice. Typically, you'd probably write your code reasonably efficient, the run a profiler and identify hotspots and only then perform these fine-grained optimizations at these particular hotspots (at least that's what I have learned over the years). This is even more important for libraries. The larger the public API the larger the maintenance burden.

> forced breaking of the rule that you should use interfaces on the left

This is fine to do in particular scenarios. For example, Guava advocates to use the Immutable collection variants wherever they are used. This rule essentially is an instance of "least required knowledge" principle. Usually, you don't need to know how exactly the map is implemented, you only need some properties of it. This is what interfaces are for. However, in some cases you might need implementation details and its perfectly fine to specialize down to the implementation.


To summarize: Yes, lambdas have some overhead. However, it doesn't matter in practically any application. Even for, e.g., programming contests, where performance is the only concern, nobody cares whether you use lambdas or not (well, to be fair, most people use C++ there ...) So just introducing new API just to avoid using lambdas is not good (IMHO!). That is what they are made for, after all. However, if there are other legitimate reasons I won't oppose this :-)

Reply:
> Technically, these two are quite different under the hood. The JVM can optimize direct method references much easier (see, e.g. here).

Interesting. Good to know. Thanks.

> Yes and no. The absolute overhead of a lambda is minimal.

If you do one or two, but lets say you are stacking these everywhere instead of normal parameters. Stuff like print(String) to printString(Supplier<String>) that calls get instantly. Or use layers of lambda after lambda (have seen 5 layers in worst case scenarios) it can have a performance impact.

One game example that I love to show like that is: Minecraft, they wen't nuts on Lambdas and started using them excessively everywhere even if it is not even reasonable to use them for their quality. The game runs since then on the server side around 50% slower then previous iterations of the game. Now lambdas are not the fault maker at here, but they are a big factor next to the design of the code in general.

> They are also useful for core performance areas, called hundreds of times per second. Simply because you save so much effort and make code maintainable so that you can spend effort on where it actually matters. Just see the example, even though there are 1 million (!) calls to the lambda, the execution takes a few milliseconds. I'd bet money that in basically every single application where the addTo method is used, a (primitive) lambda together with some optimization elsewhere has a larger impact.
Given that you have infinite time, its of course most performant to not use lambdas at all and hand tailor every single datastructure. However, usually you don't have infinite time, so prioritizing clean, readable code actually yields more performant code in practice. Same goes for the whole of the stream() API. Typically, you'd probably write your code reasonably efficient, the run a profiler and identify hotspots and only then perform these fine-grained optimizations at these particular hotspots (at least that's what I have learned over the years). This is even more important for libraries. The larger the public API the larger the maintenance burden.

Yeah I agree with you on that point, readable code leads to better performance, and lambdas can be used in high performance way even at a smaller loss but not if they are created 100k times for 1M objects to compare (ignore primitives right now and i mean 1 comparator is created per 10 objects to sort). Giving my Minecraft example.

But you can create readable code even without lambdas. It depends on your design. Lambda basically reduces the gap of skill required making readable code. Which is not a bad thing, but also not a good thing, its a thing.

> This is fine to do in particular scenarios. For example, Guava advocates to use the Immutable collection variants wherever they are used. This rule essentially is an instance of "least required knowledge" principle. Usually, you don't need to know how exactly the map is implemented, you only need some properties of it. This is what interfaces are for. However, in some cases you might need implementation details and its perfectly fine to specialize down to the implementation.

Fully agree there, just in this case that was like Collection would contain the contains method and you had to go to the next level to execute it.

> To summarize: Yes, lambdas have some overhead. However, it doesn't matter in practically any application. Even for, e.g., programming contests, where performance is the only concern, nobody cares whether you use lambdas or not (well, to be fair, most people use C++ there ...) So just introducing new API just to avoid using lambdas is not good (IMHO!). That is what they are made for, after all. However, if there are other legitimate reasons I won't oppose this :-)

"Most using C++" yeah, but the compilation process is different there but whatever.
It really depends on how often it is called. You can compare it honestly to Math.sqrt vs not to use it for distance checks.
And about not introducing a new api, it really depends. In this case this gets rid of the lazyload need in general, it results in cleaner code and faster code too since either special interfaces not being required or boxing/unboxing not being an issue.

If a better result could be achieved with a lambda for example (or lets say functional interface) then yeah go ahead, because it makes sense to get something better.

It really depends on how to add it. Which reminds me i need to make another optimization to fastutil that i noticed could save a lot of time while being almost a no change at all. You see in a minute hopefully

Reply:
Agree with most things. My main point is: Lambdas aren't bad per se, and (somehwat) excessively using them is fine for development. Once you are sure that lambdas are the problem, start to replace them.

Regarding the minecraft example: I have also worked with that code base. Many plugins were just awfully designed in terms of algorithmics (which is fine typically, because minecraft's server side doesn't have that much throughput on smaller servers). Stuff like using `LinkedList` where random access is needed, `TreeMap` where no sorting is required, using String concatenation instead of `StringBuilder`, etc. Or simply the wrong algorithmic approach to begin with. This is much more impactful than lambdas. And such issues are present in 99% of all codebases, I am very sure, especially in something accessible like minecraft.

What would be an elegant alternative here: Providing `merge` / `compute` primitive variants where it makes sense. For example `map.merge(key, value, Double::sum)`, where the last is a `DoubleBinaryOperator` (this exists in standard Java API for the "base" primitives int, long, and double). This would work for all `2Int`, `2Long`, and `2Double` maps (most other primitives don't really make sense in most cases anyway, since the JVM only really supports instructions for these three). It would also work for the "smaller" types by using "safe" casts (I think I once added similar functionality for `compute` if I recall correctly...). I'd really be interested in the performance difference of this approach vs. a direct `addTo` method. Its probably <= 20% and I am really inclined to take that penalty for the sake of flexibility (what if you'd like to instead `divideBy` or `multiplyWith`?).

> "Most using C++" yeah, but the compilation process is different there but whatever.

No? Depending on where you go, settings may be different, but, for example, ICPC uses `g++ -g -O2 -std=gnu++17 -static ${files}` and `javac -encoding UTF-8 -sourcepath . -d . ${files}`.

Reply:
> Agree with most things. My main point is: Lambdas aren't bad per se, and (somehwat) excessively using them is fine for development. Once you are sure that lambdas are the problem, start to replace them. I think we agree on that.

I can stand next to that.

> Regarding the minecraft example: I have also worked with that code base. Many plugins were just awfully designed in terms of algorithmics (which is fine typically, because minecraft's server side doesn't have that much throughput on smaller servers). Stuff like using LinkedList where random access is needed, TreeMap where no sorting is required, using String concatenation instead of StringBuilder, etc. Or simply the wrong algorithmic approach to begin with. This is much more impactful than lambdas. And such issues are present in 99% of all codebases, I am very sure, especially in something accessible like minecraft.

Interesting, yeah i can see that but if you have seen chunkloading in 1.14 or newer you will fall appart. It feels like every 3rd line of code is a new lambda on a 600 line code file. It got excessive. That's why I think its one of the reasons it runs slow. About string though. Stuff like "text"+value+"Other" text is not a bad thing in java for the whole reason: Any complex string that is created that way gets in the compilation process to byte code converted to a stringbuilder automatically. That is one of the only code changes that the compiler does that i am aware of. So seeing it in src that way means nothing. And the compiler is a good judge so far from what I have seen.

> What would be an elegant alternative here: Providing merge / compute primitive variants where it makes sense. For example map.merge(key, value, Double::sum)...........

Yeah i would be interested to see that too. Maybe i am going to write a speed test for that later.
But addTo is for one specific replacement that I have used so often that it makes sense over basic algorithmic calculation.
Before i would do this in my most optimized version:
```java
Map<Object, Integer> map = null;//Normal Hashmap

public void addStuff(Object obj, int adding)
{
	if(map.containsKey(obj))
	{
		map.put(obj, map.get(obj)+adding);
		return;
	}
	map.put(obj, adding);
}
```
That was when i was a beginner. Later it got converted to something like this:
```java
Map<Object, MutableInt> map = null;//Normal Hashmap

public void addStuff(Object obj, int adding)
{
	MutableInt value = map.get(obj);
	if(value == null)
	{
		value = new MutableInt();
		map.put(obj, value);
	}
	value.add(adding);
}
```
All that got replaced by the addToFunction because it was no longer necessary. 
It just simplifies the code so much and provides a simple mapped counter, which has so many uses.
One of the mods that I write for example is a 3D energynetwork grid that keeps track of how much energy was send over a given time. That uses the addToFunction, or shocks Entities when they collide with uninsulated cables, since it could be multiple cables it collides with at the same time it needs to combine those. Object2IntMap<Entity>,
Its rarely the case that you need multiply or division. And even if then the compute function exists to fill the gap.
But you need a mapped counter more often then anything.

> No? Depending on where you go, settings may be different, but, for example, ICPC uses g++ -g -O2 -std=gnu++17 -static ${files} and javac -encoding UTF-8 -sourcepath . -d . ${files}.

you missunderstood me here xD (my fault). I meant that lambdas in C++ are a different thing then in java. The compilation process results in 2 different results of the same named thing. Like language false friends. Like the word Die in english vs german.
In English Die means death (simplified)
In German Die is the female version of the word "The"

Reply:
> Any complex string that is created that way gets in the compilation process to byte code converted to a stringbuilder automatically.

Only if its a static string. Not if you are building a string in a loop or anything dynamic (then its a lot of copying) since the compiler then can't prove equivalence in general.

> example / replaced by addTo

Yes, and `map.merge(key, val, Integer::sum)` also does exactly that :-) `Map#merge` was introduced exactly for this type of applications! But even then, using a specialized addTo can only give you a benefit of at most x2 (instead of two hash lookups you only need one, but that's the only advantage).

This probably is faster than using `Object2IntMap` without merge or similar tricks due to Java's special handling of small Integer objects.

> Not using division often

Well, no, it just depends on the application area. If you are working with, e.g., discrete distributions, `Object2DoubleMap` is a natural representation and division / multiplication is a natural operation for, e.g., reweighting.

I just ran some JMH tests. `addTo` is only slightly faster than a `merge` with an `IntUnaryOperator` (~5% on a quick benchmark). Its also only barely faster than a "naive" approach (see below). However, all three variants manage ~80000 insertions per millisecond (!) on a single core. Note: I used `Long` as key, these have a very quick hashcode, but whatever objects you are dealing with probably has a cached hash code anyway.

Numbers:
```
Benchmark                     Mode  Cnt   Score   Error  Units
Test.runFastutilAddTo        thrpt    2  82.331          ops/s
Test.runFastutilCustomMerge  thrpt    2  79.542          ops/s
Test.runFastutilMerge        thrpt    2  63.108          ops/s
Test.runFastutilPlain        thrpt    2  77.797          ops/s
Test.runMerge                thrpt    2  53.676          ops/s
Test.runPlain                thrpt    2  53.821          ops/s
```
CustomMerge is with `IntUnaryOperator`, Merge with normal `BiFunction`, Plain with
```java
      Integer previous = map.get(key);
      if (previous == null) {
        map.put(key, value);
      } else {
        map.put(key, previous + value);
      }
```
(and similar for fastutil). 1000000 insertions per round. The difference between plain and merge becomes more apparent for more expensive `hashCode` of the key objects.

Reply:
> Only if its a static string. Not if you are building a string in a loop or anything dynamic (then its a lot of copying) since the compiler then can't prove equivalence in general. Just run something like.

Eh just made a jar with your code and this is what the bytecode pumps out. WHILE the human readable code looks like this:
ByteCode
```java
     public test() { //()V
         <localVar:index=2 , name=i , desc=I, sig=null, start=L1, end=L2>
         <localVar:index=0 , name=this , desc=Lspeiger/src/game/GameManager;, sig=null, start=L3, end=L4>
         <localVar:index=1 , name=a , desc=Ljava/lang/String;, sig=null, start=L5, end=L4>

         L3 {
             ldc "" (java.lang.String)
             astore1
         }
         L5 {
             iconst_0
             istore2
         }
         L1 {
             iload2
             ldc 1000000 (java.lang.Integer)
             if_icmpge L2
         }
         L6 {
             new java/lang/StringBuilder
             dup
             invokespecial java/lang/StringBuilder <init>(()V);
             aload1
             invokevirtual java/lang/StringBuilder append((Ljava/lang/String;)Ljava/lang/StringBuilder;);
             ldc "." (java.lang.String)
             invokevirtual java/lang/StringBuilder append((Ljava/lang/String;)Ljava/lang/StringBuilder;);
             invokevirtual java/lang/StringBuilder toString(()Ljava/lang/String;);
             astore1
         }
         L7 {
             iinc 2 1
             goto L1
         }
         L2 {
             return
         }
         L4 {
         }
     }
```
Java Code (Decompiled with fernflower):
```java
    public void test() {
        String a = "";
        for (int i = 0; i < 1000000; ++i) {
            a += ".";
        }
    }
```

The code is super un-optimized but it is converted to a stringbuilder. But my statement is still true. It is converted to a string.

But lets  give a example with a test case like this:
Java Code (Source)
```java
	public void subTest(Object key, String test, int result)
	{
		String s = "Totally not testing string builder["+key+"] functionality"+result+test;
	}
```
Java Code (Decompiled with fernflower):
```java
   public void subTest(Object key, String test, int result) {
      (new StringBuilder()).append("Totally not testing string builder[").append(key).append("] functionality").append(result).append(test).toString();
   }
```
Byte Code:
```java
     public subTest(java.lang.Object arg0, java.lang.String arg1, int arg2) { //(Ljava/lang/Object;Ljava/lang/String;I)V
         <localVar:index=0 , name=this , desc=Lspeiger/src/game/GameManager;, sig=null, start=L1, end=L2>
         <localVar:index=1 , name=key , desc=Ljava/lang/Object;, sig=null, start=L1, end=L2>
         <localVar:index=2 , name=test , desc=Ljava/lang/String;, sig=null, start=L1, end=L2>
         <localVar:index=3 , name=result , desc=I, sig=null, start=L1, end=L2>
         <localVar:index=4 , name=s , desc=Ljava/lang/String;, sig=null, start=L3, end=L2>

         L1 {
             new java/lang/StringBuilder
             dup
             invokespecial java/lang/StringBuilder <init>(()V);
             ldc "Totally not testing string builder[" (java.lang.String)
             invokevirtual java/lang/StringBuilder append((Ljava/lang/String;)Ljava/lang/StringBuilder;);
             aload1 // reference to arg0
             invokevirtual java/lang/StringBuilder append((Ljava/lang/Object;)Ljava/lang/StringBuilder;);
             ldc "] functionality" (java.lang.String)
             invokevirtual java/lang/StringBuilder append((Ljava/lang/String;)Ljava/lang/StringBuilder;);
             iload3
             invokevirtual java/lang/StringBuilder append((I)Ljava/lang/StringBuilder;);
             aload2 // reference to arg1
             invokevirtual java/lang/StringBuilder append((Ljava/lang/String;)Ljava/lang/StringBuilder;);
             invokevirtual java/lang/StringBuilder toString(()Ljava/lang/String;);
             astore4
         }
         L3 {
             return
         }
         L2 {
         }
     }
```

As you can see even the decompiler without bytecode is showing me a StringBuilder and the byte code shows it too.
If a String is shown as a String can not really be trusted in decompiler because it might as well be a stringbuilder in the actual bytecode and thats the one thats being interpreted, not the human readable source.

> I just ran some JMH tests.

Yeah roughly what i expected, but really interesting data though. I didn't expect the boxing/unboxing method to be so fast though.

Reply:
> The code is super un-optimized but it is converted to a stringbuilder. But my statement is still true. It is converted to a StringBuilder, but it results in a string.

No. The Stringbuilder is used for a single `+=` and then converted back to an (immutable) String, L6 is the loop body. Just run the code. Its orders of magnitude slower. Appending like this is `O(n^2)` while direct stringbuilder is `O(n)`.

> Yeah roughly what i expected, but really interesting data though. I didn't expect the boxing/unboxing method to be so fast though.

Integers are a special case due to the small integer cache, but boxing is indeed pretty fast (its a natural hotspot method).

Reply:
> No. The Stringbuilder is used for a single += and then converted back to an (immutable) String, L6 is the loop body. Just run the code. Its orders of magnitude slower. Appending like this is O(n^2) while direct stringbuilder is O(n).

My Point was that building a string is not always done via String.contact, but with string builders. Ofc in your scenario it uses the most optimal way to do it, but any string that is generated like how i showed it in my example is at least optimized.

> Integers are a special case due to the small integer cache, but boxing is indeed pretty fast (its a natural hotspot method).

I know about  the cache etc, but i remember when upgrading to fastutil it was much faster speed compared to the boxing/unboxing. though I think GC had a role in there too.